File: SoftwareProject-main/traffic-fine-mobile/lib/services/auth_service.dart
import 'package:firebase_auth/firebase_auth.dart';

/// NTFMS Mobile — Firebase Authentication Service
///
/// Handles officer sign-in / sign-out using Firebase Email/Password auth.
/// All officers must be registered in Firebase Console →
/// Authentication → Users before they can log in.
class AuthService {
  static final FirebaseAuth _auth = FirebaseAuth.instance;

  // ── Auth State ─────────────────────────────────────────────────────

  /// Stream of auth state changes.
  /// Emits a [User] when signed in, null when signed out.
  static Stream<User?> get authStateChanges => _auth.authStateChanges();

  /// Returns the currently signed-in Firebase User, or null.
  static User? get currentUser => _auth.currentUser;

  /// Returns true if an officer is currently signed in.
  static bool get isAuthenticated => _auth.currentUser != null;

  // ── Sign In ────────────────────────────────────────────────────────

  /// Sign in with email and password.
  ///
  /// Returns the signed-in [User] on success.
  /// Throws a user-friendly [Exception] on failure.
  static Future<User> signIn({
    required String email,
    required String password,
  }) async {
    try {
      final credential = await _auth.signInWithEmailAndPassword(
        email:    email.trim(),
        password: password,
      );
      return credential.user!;
    } on FirebaseAuthException catch (e) {
      throw Exception(_friendlyError(e.code));
    } catch (e) {
      throw Exception('Sign-in failed. Please try again.');
    }
  }

  // ── Sign Out ───────────────────────────────────────────────────────

  /// Sign out the current officer.
  static Future<void> signOut() async {
    await _auth.signOut();
  }

  // ── Password Reset ─────────────────────────────────────────────────

  /// Send a password reset email to the given address.
  static Future<void> sendPasswordReset(String email) async {
    if (email.trim().isEmpty) {
      throw Exception('Please enter your email address.');
    }
    try {
      await _auth.sendPasswordResetEmail(email: email.trim());
    } on FirebaseAuthException catch (e) {
      throw Exception(_friendlyError(e.code));
    }
  }

  // ── Error Code → Human Message ────────────────────────────────────
  static String _friendlyError(String code) {
    switch (code) {
      case 'invalid-email':
        return 'The email address is not valid.';
      case 'user-not-found':
      case 'wrong-password':
      case 'invalid-credential':
        return 'Invalid email or password. Please check your credentials.';
      case 'too-many-requests':
        return 'Too many failed attempts. Account temporarily locked. Please try again later.';
      case 'user-disabled':
        return 'This account has been disabled. Contact your administrator.';
      case 'network-request-failed':
        return 'Network error. Please check your internet connection.';
      default:
        return 'Authentication failed. Please try again.';
    }
  }
}

File: admin-portal/WebApp/src/services/authService.js
/**
 * NTFMS Admin Portal — Firebase Authentication Service
 *
 * Handles all authentication operations using Firebase Auth
 * (Email / Password provider).
 *
 * Usage:
 *   import { authService } from './authService';
 *   await authService.login(email, password);
 */

import {
  signInWithEmailAndPassword,
  signOut,
  sendPasswordResetEmail,
  onAuthStateChanged as fbOnAuthStateChanged,
} from 'firebase/auth';
import { auth } from '../firebase';

// ── User profile helper ────────────────────────────────────────────
// Maps a Firebase User object to a clean profile object used by the UI.
function buildUserProfile(firebaseUser) {
  const email    = firebaseUser.email || '';
  const username = email.split('@')[0];

  return {
    uid:      firebaseUser.uid,
    email,
    username,
    name:     firebaseUser.displayName || 'Administrator',
    role:     'Senior Police Administrator',
    district: 'Colombo Headquarters',
  };
}

// ── Auth Service ──────────────────────────────────────────────────

export const authService = {
  /**
   * Sign in with email and password.
   * @returns {{ success: boolean, user: object }}
   * @throws  Error with a human-readable message on failure
   */
  login: async (email, password) => {
    try {
      const credential = await signInWithEmailAndPassword(auth, email, password);
      const user       = buildUserProfile(credential.user);
      return { success: true, user };
    } catch (err) {
      // Convert Firebase error codes to user-friendly messages
      throw new Error(getFriendlyError(err.code));
    }
  },

  /**
   * Sign out the current user.
   */
  logout: async () => {
    await signOut(auth);
  },

  /**
   * Get the currently signed-in user profile (sync).
   * Returns null if no user is signed in.
   */
  getCurrentUser: () => {
    const u = auth.currentUser;
    return u ? buildUserProfile(u) : null;
  },

  /**
   * Returns true if a user is currently signed in.
   */
  isAuthenticated: () => !!auth.currentUser,

  /**
   * Subscribe to auth state changes.
   * Calls callback(user) whenever sign-in / sign-out happens.
   * Returns the unsubscribe function.
   *
   * @param {(user: object|null) => void} callback
   */
  onAuthStateChanged: (callback) => {
    return fbOnAuthStateChanged(auth, (firebaseUser) => {
      callback(firebaseUser ? buildUserProfile(firebaseUser) : null);
    });
  },

  /**
   * Send a password reset email.
   * @param {string} email
   */
  sendPasswordReset: async (email) => {
    if (!email || !email.includes('@')) {
      throw new Error('Please enter a valid email address.');
    }
    try {
      await sendPasswordResetEmail(auth, email);
    } catch (err) {
      throw new Error(getFriendlyError(err.code));
    }
  },
};

// ── Firebase Error Code → Human Message ──────────────────────────
function getFriendlyError(code) {
  switch (code) {
    case 'auth/invalid-email':
      return 'The email address is not valid.';
    case 'auth/user-not-found':
    case 'auth/wrong-password':
    case 'auth/invalid-credential':
      return 'Invalid email or password. Please check your credentials and try again.';
    case 'auth/too-many-requests':
      return 'Too many failed attempts. Account temporarily locked. Please try again later or reset your password.';
    case 'auth/user-disabled':
      return 'This account has been disabled. Please contact your system administrator.';
    case 'auth/network-request-failed':
      return 'Network error. Please check your internet connection and try again.';
    default:
      return 'Authentication failed. Please try again.';
  }
}
